src/zigbee2mqtt/zigbee2mqtt-device.ts

```diff
@@ -0,0 +1,261 @@
+              {
+                console.log(`Creating property for ${feature.name}`);
+
+                const property = new ColorProperty(
+                  this,
+                  'color',
+                  feature,
+                  this.client,
+                  this.deviceTopic,
+                  this.debug
+                );
+
+                this.addProperty(property);
+              }
+              break;
+          }
+        } else {
+          console.log(`Ignoring property without name: ${JSON.stringify(expose, null, 0)}`);
+        }
+      }
+    } else {
+      console.warn(`Expected features array in light expose: ${JSON.stringify(expose)}`);
+    }
+  }
+
+  private createEvents(values: string[]): void {
+    if (Array.isArray(values)) {
+      if (values.length > 0) {
+        for (const value of values) {
+          console.log(`Creating property for ${value}`);
+
+          this.addEvent(value, {
+            name: value,
+          });
+        }
+      } else {
+        console.log(`Expected object but got ${typeof values}`);
+      }
+    } else {
+      console.log(`Expected object but got ${typeof values}`);
+    }
+  }
+
+  private createAction(expose: Expos): void {
+    if (expose.name) {
+      console.log(`Creating action for ${expose.name}`);
+
+      this.addAction(expose.name, {
+        description: expose.description,
+        input: {
+          type: this.parseType(expose.type),
+          unit: expose.unit,
+          enum: expose.values,
+          minimum: expose.value_min,
+          maximum: expose.value_max,
+        },
+      });
+    } else {
+      console.log(`Ignoring action without name: ${JSON.stringify(expose, null, 0)}`);
+    }
+  }
+
+  private createProperty<T extends PropertyValue>(expose: Expos): void {
+    if (expose.name) {
+      console.log(`Creating property for ${expose.name}`);
+
+      const property = new Zigbee2MqttProperty<T>(
+        this,
+        expose.name,
+        expose,
+        this.client,
+        this.deviceTopic,
+        this.debug
+      );
+
+      this.addProperty(property);
+    } else {
+      console.log(`Ignoring property without name: ${JSON.stringify(expose, null, 0)}`);
+    }
+  }
+
+  update(update: Record<string, PropertyValue>): void {
+    if (typeof update !== 'object') {
+      console.log(`Expected object but got ${typeof update}`);
+    }
+
+    for (const [key, value] of Object.entries(update)) {
+      if (key === 'action') {
+        const event = new Event(this, value as string);
+        this.eventNotify(event);
+      } else {
+        const property = this.findProperty(key) as Zigbee2MqttProperty<PropertyValue>;
+
+        if (property) {
+          property.update(value, update);
+        } else if (this.debug) {
+          console.log(`Could not find property with name ${key}`);
+        }
+      }
+    }
+  }
+
+  performAction(action: Action): Promise<void> {
+    const { name, input } = action.asDict();
+
```

Comment:
Missing: `action.start()` and `action.finish()`

Reply:
done